src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/ChainPal.BuildChain.cs

```diff
@@ -69,21 +73,73 @@ internal sealed partial class ChainPal : IDisposable, IChainPal
-                            ChainEngine chainEngine = useMachineContext ? ChainEngine.HCCE_LOCAL_MACHINE : ChainEngine.HCCE_CURRENT_USER;
-
                             SafeX509ChainHandle chain;
                             if (!Interop.crypt32.CertGetCertificateChain(chainEngine, certificatePal.CertContext, &ft, extraStoreHandle, ref chainPara, flags, IntPtr.Zero, out chain))
                                 return null;
+                            Interop.crypt32.CertFreeCertificateChainEngine(chainEngine);
                             return new ChainPal(chain);
                         }
                     }
                 }
             }
         }
 
-        private static SafeCertStoreHandle ConvertExtraStoreToSafeHandle(X509Certificate2Collection extraStore)
+        private static IntPtr GetChainEngine(
+            X509ChainTrustMode trustMode,
+            X509Certificate2Collection customTrustStore,
+            X509Certificate2Collection extraStore,
+            bool useMachineContext
+            )
+        {
+            IntPtr chainEngine;
+            if (trustMode == X509ChainTrustMode.CustomRootTrust)
+            {
+                CERT_CHAIN_ENGINE_CONFIG customChainEngine = new CERT_CHAIN_ENGINE_CONFIG();
+                customChainEngine.cbSize = Marshal.SizeOf<CERT_CHAIN_ENGINE_CONFIG>();
+                using (SafeCertStoreHandle customTrustStoreHandle = ConvertStoreToSafeHandle(customTrustStore))
+                {
+                    if (!customTrustStoreHandle.IsInvalid)
+                    {
+                        customChainEngine.hExclusiveRoot = customTrustStoreHandle.DangerousGetHandle();
+                    }
+
+                    Interop.crypt32.CertCreateCertificateChainEngine(customChainEngine, out chainEngine);
+                    AddDefaultCertificatesToStore(extraStore, useMachineContext);
+                }
+            }
+            else
+            {
+                chainEngine = useMachineContext ? (IntPtr)ChainEngine.HCCE_LOCAL_MACHINE : (IntPtr)ChainEngine.HCCE_CURRENT_USER;
+            }
+
+            return chainEngine;
+        }
+
+        private static void AddDefaultCertificatesToStore(X509Certificate2Collection collectionStore, bool useMachineContext)
+        {
+            StoreLocation storeLocation = useMachineContext ? StoreLocation.LocalMachine: StoreLocation.CurrentUser;
+            foreach (StoreName storeName in (StoreName[]) Enum.GetValues(typeof(StoreName)))
+            {
+                if (storeName == StoreName.Disallowed)
+                {
+                    continue;
+                }
+
+                X509Store store = new X509Store(storeName, storeLocation);
+                store.Open(OpenFlags.ReadOnly);
+
+                foreach (X509Certificate2 certificate in store.Certificates)
+                {
+                    collectionStore.Add(certificate);
```

Comment:
I am adding the default store certificates to extra store. From MSDN 
> Exclusive trust mode allows applications to specify trust anchors and peer-trusted certificates for certificate chain validation. In the exclusive trust mode, the root store and the trusted people store on the system are ignored, and the anchors and certificates pointed to by the hExclusiveRoot and hExclusiveTrustedPeople members are used instead.

To avoid a partial chain I need to add the root store and trusted people store to the extraStore collection so that they can be passed in as untrusted. I will add a comment about this to prevent confusion.

Reply:
Ah, somehow I read that backwards (that you were adding to the stores, not draining the stores).

I'd rather that this didn't do this.  If Windows isn't looking at the default stores then we shouldn't either (and it would remove variations of this code from all three platforms).  I expect 99% of the use cases to be "if it isn't chained to this/one-of-these root(s) I don't want it", so PartialChain vs UntrustedRoot wouldn't matter much.

Reply:
Ok that makes things simpler. The one issue is that on osx I couldn't find a way not to check the default stores. Setting the trust anchor certificates only affects which certificates are trusted and not what stores are searched.

So Windows would produce a PartialChain and OSX UnstrustedRoot.